hibernate-core/src/main/java/org/hibernate/hql/spi/id/cte/AbstractCteValuesListBasedBulkIdHandler.java

```diff
@@ -0,0 +1,150 @@
+		this.catalog = catalog;
+		this.schema = schema;
+
+		final AbstractRestrictableStatement statement = (AbstractRestrictableStatement) walker.getAST();
+		final FromElement fromElement = statement.getFromClause().getFromElement();
+
+		this.targetedPersister = fromElement.getQueryable();
+
+		final ProcessedWhereClause processedWhereClause = processWhereClause( statement.getWhereClause() );
+		this.idSelectParameterSpecifications = processedWhereClause.getIdSelectParameterSpecifications();
+
+		final String bulkTargetAlias = fromElement.getTableAlias();
+
+		this.idSelect = generateIdSelect( bulkTargetAlias, processedWhereClause ).toStatementString();
+	}
+
+	@Override
+	public Queryable getTargetedQueryable() {
+		return targetedPersister;
+	}
+
+	protected JDBCException convert(
+			SQLException e,
+			String message,
+			String sql) {
+		throw factory().getSQLExceptionHelper().convert( e, message, sql );
+	}
+
+	protected String determineIdTableName(Queryable persister) {
+		return jdbcEnvironment.getQualifiedObjectNameFormatter().format(
+				new QualifiedTableName(
+						Identifier.toIdentifier( catalog ),
+						Identifier.toIdentifier( schema ),
+						Identifier.toIdentifier( "HT_" + persister.getTableName() )
+				),
+				jdbcEnvironment.getDialect()
+		);
+	}
+
+	protected String generateIdSubselect(Queryable persister) {
+		return new StringBuilder()
+			.append( "select " )
+			.append( StringHelper.join( ", ", persister.getIdentifierColumnNames() ) )
+			.append( " from " )
+			.append( determineIdTableName( persister ) )
+			.toString();
+	}
+
+	protected CteValuesListBasedBuilder prepareCteStatement(
+			SharedSessionContractImplementor session,
+			QueryParameters queryParameters) {
+
+		try {
+			try (PreparedStatement ps = session.getJdbcCoordinator()
+					.getStatementPreparer()
+					.prepareStatement( idSelect, false )) {
+				int sum = 1;
+				for ( ParameterSpecification parameterSpecification : idSelectParameterSpecifications ) {
+					sum += parameterSpecification.bind( ps, queryParameters, session, sum );
+				}
+				List<Object[]> ids = new ArrayList<>();
+
+				ResultSet rs = session
+						.getJdbcCoordinator()
+						.getResultSetReturn()
+						.extract( ps );
+				while ( rs.next() ) {
+					Object[] result = new Object[targetedPersister.getIdentifierColumnNames().length];
+					for ( String columnName : targetedPersister.getIdentifierColumnNames() ) {
+						Object column = rs.getObject( columnName );
+						result[rs.findColumn( columnName ) - 1] = column;
+					}
+					ids.add( result );
+				}
+
+				if ( ids.isEmpty() ) {
+					return null;
+				}
+
+				return new CteValuesListBasedBuilder(
+						determineIdTableName( targetedPersister ), targetedPersister.getIdentifierColumnNames(), ids
+				);
+			}
+		}
+		catch ( SQLException e ) {
+			throw convert( e, "could not insert/select ids for bulk operation", idSelect );
```

Comment:
Good catch. I changed it to `could not select ids for bulk operation` since we only select the ids in this phase.